vllm/v1/worker/gpu_model_runner.py

```diff
@@ -832,6 +900,44 @@ def make_sampling_metadata(
                 self.top_p_cpu_tensor[:self.num_reqs], non_blocking=True)
             self.top_k[:self.num_reqs].copy_(
                 self.top_k_cpu_tensor[:self.num_reqs], non_blocking=True)
+            if not self.no_reqs_with_penalties:
+                # Since syncing these tensors is expensive only copy them
+                # if necessary i.e. if there are requests which require
+                # penalties to be applied during sampling.
+                self.frequency_penalties[:self.num_reqs].copy_(
+                    self.frequency_penalties_cpu_tensor[:self.num_reqs],
+                    non_blocking=True)
+                self.presence_penalties[:self.num_reqs].copy_(
+                    self.presence_penalties_cpu_tensor[:self.num_reqs],
+                    non_blocking=True)
+                self.repetition_penalties[:self.num_reqs].copy_(
+                    self.repetition_penalties_cpu_tensor[:self.num_reqs],
+                    non_blocking=True)
+                # The prompt tokens are used only for applying penalties during
+                # the sampling process. Hence copy these tensors only when
+                # there are requests which need penalties to be applied.
+                self.prompt_tokens_tensor = \
+                    self._construct_prompt_tokens_tensor(
+                        requests, self.vocab_size, device=self.device)
+
+        output_token_ids: List[List[int]] = []
+        min_tokens: List[int] = []
+        stop_token_ids: List[set[int]] = []
```

Comment:
Why don't we cache these values in `InputBatch` rather than recreating them every time?

Reply:
Done